### ISSUE TITLE
[SET_CERTIFICATE] Save certificate into SPDM context

### ISSUE THREAD

**Issue by rw8896** [DMTF/libspdm#873]
https://github.com/DMTF/libspdm/blob/a27fe8164d8edcbe3143e5b9e65f0c378c2df77a/library/spdm_responder_lib/libspdm_rsp_set_certificate.c#L138-L147

If reset is not required, should the certificate chain be added into spdm_context?
Note: in case of AliasCert model, it might need to extract Alias cert from slot 0 and append the Alias cert to compose a complete certificate chain.

**Comment by jyao1** [comment]
DeviceCert mode + reset needed: Not save to context
AliasCert mode + reset needed: Not save to context

DeviceCert mode + reset not needed: Can save to context <==
AliasCert mode + reset not needed: Not sure if it is valid? 

Question: libspdm save the context? or Integrator save the context ?
Conclusion: It is easy that libspdm save the context.
We can let integrator save to the context by using SetData(), because the integrator can know the slot id and full cert chain data. But libspdm_write_certificate_to_nvm() does not have libspdm_context. It seems problematic. Maybe define a callback function?

**Comment by rw8896** [comment]
Integrator saves Certificates into context seems a better solution.
- libspdm_write_certificate_to_nvm() didn't have libspdm_context: could we pass spdm context to the function?
Anyways, the integrator should have that context. Otherwise the integrator won't be able to set Certificates during libspdm initialization.
- current libspdm_write_certificate_to_nvm() returns bool, could we change it to return a status?
If integrator can't update the full Certificates into the spdm context, it should return SPDM_ERROR_CODE_RESET_REQUIRED or a specific error code. spdm_context->need_reset_to_set_cert is not necessary then.

**Comment by jyao1** [comment]
Another special corner case, that a device may setup 2 spdm_context, such as PCI_DOE and MCTP.

If a certificate chain is setup from one context, do we need provision to the other context ?

**Comment by jyao1** [comment]
> Integrator saves Certificates into context seems a better solution.
> 
> * libspdm_write_certificate_to_nvm() didn't have libspdm_context: could we pass spdm context to the function?
>   Anyways, the integrator should have that context. Otherwise the integrator won't be able to set Certificates during libspdm initialization.
> * current libspdm_write_certificate_to_nvm() returns bool, could we change it to return a status?
>   If integrator can't update the full Certificates into the spdm context, it should return SPDM_ERROR_CODE_RESET_REQUIRED or a specific error code. spdm_context->need_reset_to_set_cert is not necessary then.

**Comment by jyao1** [comment]
> Integrator saves Certificates into context seems a better solution.
> 
> * libspdm_write_certificate_to_nvm() didn't have libspdm_context: could we pass spdm context to the function?
>   Anyways, the integrator should have that context. Otherwise the integrator won't be able to set Certificates during libspdm initialization.

No. We discussed in the meeting. The device secrete lib should be context independent.

> * current libspdm_write_certificate_to_nvm() returns bool, could we change it to return a status?
>   If integrator can't update the full Certificates into the spdm context, it should return SPDM_ERROR_CODE_RESET_REQUIRED or a specific error code. spdm_context->need_reset_to_set_cert is not necessary then.

No. We discussed in PR. need_reset should be a predefined attribute from the integrator. No need to return every time.

**Comment by steven-bellock** [comment]
There probably needs to be two callbacks to the integrator. One for authorization and the other to populate the spdm context with the new certificate.

**Comment by rw8896** [comment]
> > Integrator saves Certificates into context seems a better solution.
> > 
> > * libspdm_write_certificate_to_nvm() didn't have libspdm_context: could we pass spdm context to the function?
> >   Anyways, the integrator should have that context. Otherwise the integrator won't be able to set Certificates during libspdm initialization.
> 
> No. We discussed in the meeting. The device secrete lib should be context independent.
> 

Would you elaborate more about " The device secrete lib should be context independent"?
I am okay with the principle but just curious what's the design consideration here.

Back to set certificate, why is this function in device secret lib?
Measurements and keys are device secret, but saving certificates into NV sounds more like a general callback function the integrator could register?

**Comment by steven-bellock** [comment]
> Would you elaborate more about " The device secrete lib should be context independent"?
I am okay with the principle but just curious what's the design consideration here.

non-libspdm modules that are provided by the integrator may be realized entirely in hardware, so it would be weird to pass the spdm_context to them. It provides a boundary between libspdm and the modules it depends on, and maintains the spdm_context as an opaque structure.

>Back to set certificate, why is this function in device secret lib?
Now that its scope has increased "secret lib" may not be the appropriate name anymore. It's more like "subsystem within a Responder that provides confidentiality and integrity".

**Comment by rw8896** [comment]
> > Would you elaborate more about " The device secrete lib should be context independent"?
> > I am okay with the principle but just curious what's the design consideration here.
> 
> non-libspdm modules that are provided by the integrator may be realized entirely in hardware, so it would be weird to pass the spdm_context to them. It provides a boundary between libspdm and the modules it depends on, and maintains the spdm_context as an opaque structure.
> 

For this purpose actually we could use void * in the callback function, e.g. 
`void spdm_server_connection_state_callback(
    void *spdm_context, libspdm_connection_state_t connection_state)`

Current device secret lib functions doesn't need the spdm_context but when it's necessary, using void * doesn't seem to break the boundary to maintain spdm_context as an opaque structure.

Usually the integrator should already implement the code to compose/verify full cert chain (from provisioned certs) and then set data into spdm_context, e.g. https://github.com/DMTF/spdm-emu/blob/e85b78f5e0daec0632c5c6dfbf79d22548bc6cf2/spdm_emu/spdm_responder_emu/spdm_responder_spdm.c#L240-L241

It would make more sense to leave saving certchain into context for the integrator and allow integrator to decide reset is needed or not at run-time.
